subsys/fs/nvs/nvs.c

```diff
@@ -6,104 +6,492 @@
-int _nvs_sector_hdr_get(struct nvs_fs *fs, off_t offset,
-			struct _nvs_sector_hdr *sector_hdr)
+/* basic flash erase */
+int _nvs_flash_erase(struct nvs_fs *fs, off_t offset, size_t len)
 {
-	return flash_read(fs->flash_device,
-			  fs->offset + (offset & ~(fs->sector_size - 1)),
-			  sector_hdr,
-			  sizeof(*sector_hdr));
+	int rc;
+
+	rc = flash_write_protection_set(fs->flash_device, 0);
+	if (rc) {
+		/* flash protection set error */
+		return rc;
+	}
+	SYS_LOG_DBG("Erasing flash at %"PRIx32", len %d", offset, len);
+	rc = flash_erase(fs->flash_device, fs->offset + offset, len);
+	if (rc) {
+		/* flash write error */
+		return rc;
+	}
+	(void) flash_write_protection_set(fs->flash_device, 1);
+	return 0;
 }
 
-/* Initializing sector by writing magic and status to sector header,
- * set fs->write_address just after header offset should be pointing
- * to the sector
+/* end of basic flash routines */
+
+/* advanced flash routines */
+
+/* _nvs_flash_block_cmp compares the data in flash at offset to data
+ * in blocks of size NVS_BLOCK_SIZE, returns 0 if equal
  */
-int _nvs_sector_init(struct nvs_fs *fs, off_t offset)
+int _nvs_flash_block_cmp(struct nvs_fs *fs, off_t offset, const void *data,
+		   size_t len)
+{
+	int rc, block_size;
+	u8_t buf[max(NVS_BLOCK_SIZE, fs->write_block_size)];
+
+	block_size = sizeof(buf);
+
+	while (len >= block_size) {
+		rc = _nvs_flash_read(fs, offset, buf, block_size);
+		if (rc) {
+			return rc;
+		}
+		rc = memcmp(data, buf, block_size);
+		if (rc) {
+			return rc;
+		}
+		len -= block_size;
+		offset += block_size;
+		data += block_size;
+	}
+	if (len) {
+		rc = _nvs_flash_read(fs, offset, buf, len);
+		if (rc) {
+			return rc;
+		}
+		rc = memcmp(data, buf, len);
+		if (rc) {
+			return rc;
+		}
+	}
+	return 0;
+}
+
+/* flash block move: move a block at offset to the current write location
+ * and updates the write location, len should be multiple of
+ * fs->write_block_size. _nvs_flash_move does a verification of the copy.
+ */
+int _nvs_flash_block_move(struct nvs_fs *fs, off_t offset, size_t len)
+{
+	int rc, byte_to_copy, block_size;
+	u8_t buf[max(NVS_BLOCK_SIZE, fs->write_block_size)];
+	off_t write_location;
+
+	write_location = fs->write_location;
+
+	block_size = sizeof(buf);
+
+	while (len > 0) {
+		byte_to_copy =	min(block_size, len);
+		rc = _nvs_flash_read(fs, offset, buf, byte_to_copy);
+
+		if (rc) {
+			return rc;
+		}
+		rc = _nvs_flash_write(fs, write_location, buf,
+				      byte_to_copy);
+		if (rc) {
+			return rc;
+		}
+		rc = _nvs_flash_block_cmp(fs, write_location, buf,
+					  byte_to_copy);
+		if (rc) {
+			return rc;
+		}
+		len -= byte_to_copy;
+		offset += byte_to_copy;
+		write_location += byte_to_copy;
+	}
+	/* update fs->write_location when all writes are successful */
+
+	fs->write_location = write_location;
+
+	return 0;
+}
+
+/* flash write and verify */
+int _nvs_flash_write_ver(struct nvs_fs *fs, off_t offset, const void *data,
```

Comment:
I know that `ver` means "verify", but "version" first came to my mind. I think it would make it more understandable to call it `_nvs_flash_write_cmp`, and would be consistent with `_nvs_flash_block_cmp`.

Reply:
Yes, _cmp is more consistent, will change